### PR TITLE
[stdlib][NFC] Put stored properties & primary initializers before other members in all struct declarations

### PR DESCRIPTION
**Pull request by lorentey**
Putting stored properties and critical initializers at the top makes it easier to understand these definitions. (I kept having to scroll/search around to find these, which gets rather tiresome after a couple of cases.)

Also, add a section to the stdlib programmers manual explaining this preference.  And while we're updating that file, let's also expand on line breaking conventions a bit.

This PR also includes a small number of trivial formatting fixes (removing extra indentation, not letting doc comments separate function attributes from their declaration, etc.)

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Review by Azoy — APPROVED** [review]
This is a very nice change for the stdlib. Thank you so much Karoy!

**Review by milseman — APPROVED** [review]
Overall LGTM. We'll want to figure out when to merge this so as to reduce the amount of conflicts in already open PRs.

**Review by glessard — APPROVED** [review]
Things are clearly more readable when re-ordered in this way.

File: docs/StandardLibraryProgrammersManual.md [hunk 1]
```diff
@@ -12,10 +12,136 @@ In this document, "stdlib" refers to the core standard library (`stdlib/public/c
...
+
+#### Line Breaking
+
+The stdlib currently has a hard line length limit of 80 characters. This allows code to be easily read in environments that don't gracefully handle long lines, including (especially!) code reviews on GitHub.
    
 We use two spaces as the unit of indentation. We don't use tabs.
 
+To break long lines, please closely follow the indentation conventions you see in the existing codebase. (FIXME: Describe in detail.)
+
+Our primary rule is that if we need to break a list (such as arguments, tuple or array/dictionary literals, generic type parameters, etc.), then we always put each item on its own line, indented by +1 unit, even if a series of items would fit on a single line together.
```
[the remark this answers is not in the document]
Comment (lorentey) [reply docs/StandardLibraryProgrammersManual.md:25]: Oh that's a good point. I will adjust the wording.

File: stdlib/public/core/Substring.swift [hunk 2]
```diff
@@ -343,14 +351,6 @@ extension Substring.UTF8View: BidirectionalCollection {
...
   public typealias SubSequence = Substring.UTF8View
 
-  /// Creates an instance that slices `base` at `_bounds`.
-  @inlinable
-  internal init(_ base: String.UTF8View, _bounds: Range<Index>) {
-    _slice = Slice(
-      base: String(base._guts).utf8,
-      bounds: _bounds)
-  }
-
```
[the remark this answers is not in the document]
Comment (lorentey) [reply stdlib/public/core/Substring.swift:353]: I think this change has the side effect of suppressing the compiler-provided internal memberwise initializer (which feels like a good thing, as that initializer is never used).

**Comment by lorentey** [comment]
@swift-ci smoke test

File: docs/StandardLibraryProgrammersManual.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lorentey) [reply docs/StandardLibraryProgrammersManual.md]: (I also added an explicit example for breaking a list only at the delimiters.)

File: docs/StandardLibraryProgrammersManual.md [hunk 3]
```diff
@@ -22,24 +24,24 @@ We use two spaces as the unit of indentation. We don't use tabs.
 
 To break long lines, please closely follow the indentation conventions you see in the existing codebase. (FIXME: Describe in detail.)
 
-Our primary rule is that if we need to break a list (such as arguments, tuple or array/dictionary literals, generic type parameters, etc.), then we always put each item on its own line, indented by +1 unit, even if a series of items would fit on a single line together. 
+Our primary rule is that if we need to put a line break anywhere in the middle of a list (such as arguments, tuple or array/dictionary literals, generic type parameters, etc.), then we need to go the full way, and put each item on its own line, indented by +1 unit, even if some of the items would fit on a single line together.
```
Comment (glessard) [inline docs/StandardLibraryProgrammersManual.md:27]: Nit:
```suggestion
Our primary rule is that if we need to put a line break anywhere in the middle of a list (such as arguments, tuple or array/dictionary literals, generic type parameters, etc.), then we should go all the way, and put each item on its own line, indented by +1 unit, even if some of the items would fit on a single line together.
```

Comment (glessard) [reply docs/StandardLibraryProgrammersManual.md:27]: Another one:
```suggestion
Our primary rule is that if we need to insert a line break anywhere in the middle of a list (such as arguments, tuple or array/dictionary literals, generic type parameters, etc.), then we need to go the full way, and put each item on its own line, indented by +1 unit, even if some of the items would fit on a single line together.
```

Comment (lorentey) [reply docs/StandardLibraryProgrammersManual.md:27]: I sneakily edited your first suggestion to change "then we need to go" to "then we must go" (RFC 2119 style), to emphasize that this is not something people can opt out of if they don't like it. 🤓

File: docs/StandardLibraryProgrammersManual.md [hunk 4]
```diff
@@ -12,10 +12,232 @@ In this document, "stdlib" refers to the core standard library (`stdlib/public/c
...
+
+- In some cases, it can also work well to declare the most essential protocol conformances directly on the type definition; feel free to do so if it helps understanding. (You can still implement requirements in separate extensions in this case, or you can do it within the main declaration.)
+
+- It's okay for the core type declaration to forward reference large nested types or static members that are defined in subsequent extensions. It's often a good idea to define these in an extension immediately following the type declaration, but this is not a strict rule. 
+
+Extensions are a nice way to break up the implementation into easily digestable chunks, but they aren't the only way. The goal is to make things easy to understand -- if a type is small enough, it may be best to list every member directly in the `struct`/`class` definition, while for huge types it often makes more sense to break them up into number of separate source files. 
+
+```
```
Comment (lorentey) [inline docs/StandardLibraryProgrammersManual.md:177]:
```suggestion
```swift
```

**Pull request merged**